Explore: Show ANSI colors when highlighting matched words in the logs panel

oliverfrye opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

Currently, ANSI color codes are rendered in the Explore view for Loki logs. However, if a line filter is used (e.g. `|= "info"`) then the ANSI color codes are no longer rendered, and are instead displayed in their escaped form (e.g. `[36m`) in the log line. This is because a `Highlighter` component is used to highlight the searched for word in the output, which is not ANSI-aware.

![image](https://user-images.githubusercontent.com/11541660/139013723-bad01209-fcb3-440d-9457-fe060a0252d9.png)

This PR modifies the existing ANSI-aware `LogMessageAnsi` component to be capable of highlighting the searched for word (still using the `Highlighter` component) in addition to the ANSI coloring.

![image](https://user-images.githubusercontent.com/11541660/139013793-30c9f90d-d387-4d5e-a434-50b7e34704a7.png)


**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #40265 

**Special notes for your reviewer**:

This is my first Grafana PR, and first time working with React. I'm not sure if my use of multiple `Highlighter` components is problematic in terms of performance - my intention was to avoid having to re-implement its functionality inside `LogMessageAnsi`.

A known issue: highlighting won't work across chunks (blocks of text separated by ANSI escape characters, as I understand) because the `Highlighter` component wraps each `chunk` individually. I don't think this is actually a problem in practice, however, because Loki doesn't support matching line filters that have substrings containing escape codes (see https://github.com/grafana/loki/issues/2861 for more details)

oliverfrye committed:
Explore: Show ANSI colors when highlighting matched words

gabor commented:
hi @oliverfrye , thank you for the pull request. could you please sign the CLA as mentioned in https://github.com/grafana/grafana/pull/40971#issuecomment-952595122 ? thanks.

oliverfrye commented:
> hi @oliverfrye , thank you for the pull request. could you please sign the CLA as mentioned in [#40971 (comment)](https://github.com/grafana/grafana/pull/40971#issuecomment-952595122) ? thanks.

Hi @gabor, my apologies for the delay. I'm still awaiting review from my employer's legal team before I am permitted to sign the CLA. I will work on having it signed as soon as possible.

gabor commented:
> > hi @oliverfrye , thank you for the pull request. could you please sign the CLA as mentioned in [#40971 (comment)](https://github.com/grafana/grafana/pull/40971#issuecomment-952595122) ? thanks.
> 
> Hi @gabor, my apologies for the delay. I'm still awaiting review from my employer's legal team before I am permitted to sign the CLA. I will work on having it signed as soon as possible.

hi @oliverfrye ,sure, no problem.  👍

oliverfrye commented:
Hi @gabor, I got approval from my company, so the CLA is now signed and this is ready for review. Thank you for your patience 🙏

gabor commented:
hi @oliverfrye , the PR looks good, it is working fine, good job!

the only thing i am still investigating is backward-compatibility... you see, we export `LogMessageAnsi` from `@grafana/ui`, so any change to it's `Props` can be a backward-incompatible change to `@grafana/ui` , so we have to change it's props very carefully...

currently those extra props (highlight related) basically re-export props from `react-highlight-words`. this can be problematic, you see, if later we decide to upgrade to a new version of `react-highlight-words`, it might cause a backward-incompatibility in `LogMessageAnsi`.

for this reason, i'd prefer to modify `LogMessageAnsi`'s Props to be roughly like this:
```typescript
interface Props {
  value: string;
  highlight?: {
    searchWords: string[];
    highlightClassName: string;
  },
};
```

we store all highlight-related variables in it's own structure, and we make the structure optional, so this will be backward-compatible with old code.

also, we do not have to send in the `findChunks` function, we can just import it from `@grafana/data`, the way `LogRowMessage.tsx` does it. this is important, because that function has a very large type-signature, and this way we are not publishing that in the public API.

there are two ways to go from here:
1. you could do  this change in this PR. or,
2. we can merge your PR as it is, in it's current form, and i will fix the Props API in a separate pull request.

which one do you prefer?

oliverfrye commented:
@gabor that's excellent feedback, thank you! Your suggestions make it much cleaner.

I've implemented the requested changes and squashed / rebased my PR to include them.

gabor commented:
hi @oliverfrye , thanks for the changes. it seems the typechecking failed on your changes, the reason is that you need a `@ts-ignore` above the react-highlight import, like it is at https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/components/Logs/LogRowMessage.tsx#L7-L8

(this ts-ignore solution is not optimal, but that's what the other places in grafana-ui do too, so for now let's go with that 😄 )

oliverfrye commented:
@gabor ah, what a shame. The type error didn't show up in my error, which was unexpected. Nevertheless, I've added `@ts-ignore` like it is in `LogRowMessage.tsx`.

gabor commented on packages/grafana-ui/src/components/Logs/LogRowMessage.tsx:
```diff
@@ -63,17 +63,18 @@ function renderLogMessage(
-  if (needsHighlighter) {
+  const searchWords = highlights ?? [];
+  if (hasAnsi) {
+    return <LogMessageAnsi value={entry} highlight={{ searchWords, highlightClassName }} />;
```
i think we should use the `needsHighlighter` boolean in the ansi-highlight case too, to be in sync with the other branches, something like this:

```suggestion
    const highlight = needsHighlighter ? {searchWords, highlightClassName} : undefined;
    return <LogMessageAnsi value={entry} highlight={highlight} />;
```

gabor reviewed (changes requested):
nice work 👍 ! it works fine, and the code is basically good too, except one minor issue i described below...

we should only allow ansi-highlighting when `needsHighlighter` is `true`, there are some checks in that boolean, like max-string-length etc.

oliverfrye committed:
only highlight ANSI text if needsHighlighter is true

Co-authored-by: Gábor Farkas <gabor.farkas@gmail.com>

oliverfrye commented:
Good call, thanks 👍

gabor reviewed (approved):
works nicely!

gabor commented on packages/grafana-ui/src/components/Logs/LogRowMessage.tsx:
```diff
@@ -63,17 +63,19 @@ function renderLogMessage(
-  if (needsHighlighter) {
+  const searchWords = highlights ?? [];
+  if (hasAnsi) {
+    const highlight = needsHighlighter ? {searchWords, highlightClassName} : undefined;
```
i'm sorry, the change i requested made the linter angry 😢 , i think we need to put spaces inside the curly-braces... to be sure, please run `yarn lint`, then fix the problem and run `yarn lint` again to be sure the problems is solved. thanks!

oliverfrye replied:
OK, no problem, should be sorted now.

gabor reviewed (changes requested):
sorry, there is a linter error that needs to be fixed before we merge this. otherwise everything is ok.

oliverfrye committed:
fix lint error

gabor reviewed (approved):
LGTM'

gabor commented:
@oliverfrye thanks for the pull-request, it has been merged!
